### PR TITLE
Class simplifications

### PR DESCRIPTION
**Pull request by EmilyBourne**
This PR simplifies the handling of classes in a few locations. This will notably be useful for #1607 .

**Commit summary**
- Add a `bound_argument` property to `FunctionDefArgument` to identify class arguments that are bound to the calling function
- Add a `class_type` property to `ClassDef` to store the associated `DataType`
- Remove `DottedFunctionCall`. A separate concept is not needed for this.
- Use `FunctionDef.cls_name` to save the alternative name of a class method.
- Rename `FunctionDef` objects which are printed with alternative names to ensure that these names are collisionless and can be accessed elsewhere in Pyccel (e.g. in the wrapper)
- Remove dead code

### CODE REVIEW

File: pyccel/codegen/printing/fcode.py [hunk 1]
```diff
@@ -449,14 +474,16 @@ def _print_Module(self, expr):
             name = self._print(expr.name)
         name = name.replace('.', '_')
         if not name.startswith('mod_') and self.prefix_module:
-            name = '{prefix}_{name}'.format(prefix=self.prefix_module,
-                                            name=name)
+            name = f'{self.prefix_module}_{name}'
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:477]: /bot accept
See #1677

File: pyccel/codegen/printing/fcode.py [hunk 2]
```diff
@@ -437,6 +437,31 @@ def _get_external_declarations(self):
...
+                if m_name in _default_methods:
+                    suggested_name = _default_methods[m_name]
+                    scope.rename_function(method, suggested_name)
+                method.cls_name = scope.get_new_name(f'{name}_{method.name}')
+        for i in expr.interfaces:
+            for f in i.functions:
+                if not f.is_inline:
+                    i_name = f.name
+                    if i_name in _default_methods:
+                        suggested_name = _default_methods[i_name]
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/codegen/printing/fcode.py:461]: /bot accept
Currently only `create` and `free` are found in `_default_methods`. These do not appear in interfaces. However this code should be kept to support magic methods in the future

**Review by yguclu — CHANGES_REQUESTED** [review]
Looks good to me!

File: pyccel/ast/core.py [hunk 3]
```diff
@@ -1940,6 +1945,21 @@ def make_const(self):
...
+
+        Indicate if the argument is bound to the function call. This is
+        the case if the argument is the first argument of a method of a
+        class.
+        """
+        return self._bound_argument
+
+    @bound_argument.setter
+    def bound_argument(self, bound):
+        self._bound_argument = bound
```
[the remark this answers is not in the document]
Comment (EmilyBourne) [reply pyccel/ast/core.py]: Fixed bc40a292b

**Review by yguclu — APPROVED** [review]

**Review by yguclu — APPROVED** [review]

**Pull request merged**